index.js

```diff
@@ -2429,147 +2371,185 @@
-    impl          // :: Function
+  //  withTypeChecking :: ... -> Function
+  function withTypeChecking(
+    env,            // :: Array Type
+    typeInfo,       // :: TypeInfo
+    impl            // :: Function
   ) {
     var n = typeInfo.types.length - 1;
 
-    var curried = arity(_indexes.length, function() {
-      if (opts.checkTypes) {
-        var delta = _indexes.length - arguments.length;
-        if (delta < 0) throw tooManyArguments(typeInfo, n - delta);
+    //  wrapFunctionCond :: (TypeVarMap, Integer, a) -> a
+    function wrapFunctionCond(_typeVarMap, index, value) {
+      if (typeInfo.types[index].type !== FUNCTION) return value;
+
+      var expType = typeInfo.types[index];
+      var isValid = test(env);
+
+      //  checkValue :: (TypeVarMap, Integer, String, a) -> Either (() -> Error) TypeVarMap
+      function checkValue(typeVarMap, index, k, x) {
+        var propPath = [k];
+        var t = expType.types[k].type;
+        return (
+          t.type === VARIABLE ?
+            (function(typeVarMap) {
+               return isEmpty(typeVarMap[t.name].types) ?
+                 Left(function() {
+                   return typeVarConstraintViolation(
+                     env,
+                     typeInfo,
+                     index,
+                     propPath,
+                     typeVarMap[t.name].valuesByPath
+                   );
+                 }) :
+                 Right(typeVarMap);
+             }(updateTypeVarMap(env, typeVarMap, t, index, propPath, [x]))) :
+          isValid(t)(x) ?
+            Right(typeVarMap) :
+          // else
+            Left(function() {
+              return invalidValue(env, typeInfo, index, propPath, x);
+            })
+        );
       }
+
+      var isThunk = expType.types.$1.type.type === NO_ARGUMENTS;
+      var numArgsExpected = isThunk ? 0 : expType.keys.length - 1;
       var typeVarMap = _typeVarMap;
-      var values = _values.slice();
-      var indexes = [];
-      for (var idx = 0; idx < _indexes.length; idx += 1) {
-        var index = _indexes[idx];
-
-        if (idx < arguments.length &&
-            !(typeof arguments[idx] === 'object' &&
-              arguments[idx] != null &&
-              arguments[idx]['@@functional/placeholder'] === true)) {
-
-          var value = arguments[idx];
-          if (opts.checkTypes) {
-            var result = satisfactoryTypes(opts.env,
-                                           typeInfo,
-                                           typeVarMap,
-                                           typeInfo.types[index],
-                                           index,
-                                           [],
-                                           [value]);
-            typeVarMap = assertRight(result).typeVarMap;
-          }
-          values[index] = value;
-        } else {
-          indexes.push(index);
+      return function(x) {
+        var args = slice.call(arguments);
```

Comment:
I think we might be able to avoid calling `slice.call` here. It would increase performance.

Reply:
Great point!

:zap: